### ISSUE TITLE
https://github.com/ultralytics/ultralytics/issues/22056,There are still issues.

### ISSUE THREAD

**Issue by JL-52Hertz** [ultralytics/ultralytics#22059]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I was too happy too early; the issue with 22056 is still unresolved. If I simply put the data loading outside of if RANK{-1,0}, the program can run, but the logic is wrong. However, if I put all the pruning logic and data loading outside, it is incorrect because my scoring function rewrites the validator class. So the current issues are still unresolved!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JL-52Hertz, thank you for your report and for continuing to dig into this 🚀 This is an automated response to help get things moving quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples that may help clarify data loading, pruning, and custom Validator workflows.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Given your notes about data loading scope (RANK{-1,0}), pruning logic placement, and a custom validator/scoring function, please include:
- A short script showing exactly how you’re placing data loading, pruning, and your custom Validator subclass
- The command you run (CLI or Python) and logs/output
- Environment details (OS, Python, PyTorch, CUDA) and ultralytics version
- Whether you’re running single-GPU, multi-GPU, or DDP, and your world size
- A tiny sample dataset or synthetic data stub if applicable

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Screenshots or snippets that show how the custom validator interacts with pruning will be especially helpful.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Once you share the MRE and details above, we’ll be able to reproduce and investigate your 22056-related issue more accurately. Thank you for your patience and thoroughness 🙏

**Comment by JL-52Hertz** [comment]
@glenn-jocher Please take a look at this issue https://github.com/ultralytics/ultralytics/issues/22056, looking forward to your reply, thank you!

**Comment by Y-T-G** [comment]
Only validation loader should be inside the if block. The train loader should created outside.

**Comment by JL-52Hertz** [comment]
1

**Comment by glenn-jocher** [comment]
Thanks for the details—those “Scanning … train.cache” lines indicate you’re rebuilding the train dataset each epoch; keep the train dataloader built once per process and only run your prune scoring on rank 0, then broadcast the keep-indices and update the existing dataset in place before resetting the iterator, e.g.:

```python
import torch.distributed as dist
import numpy as np

# rank 0: score + decide which samples to keep
keep = compute_keep_indices(...)  # your validator/scoring
obj = [keep]
if dist.is_initialized():
    dist.broadcast_object_list(obj, src=0)  # share to all ranks
keep = obj[0]

# all ranks: update train dataset without rebuilding it
ds = trainer.train_loader.dataset  # your Trainer instance
ds.indices = np.asarray(keep, dtype=np.int64)  # adjust to your dataset field
trainer.train_loader.reset()  # reuse workers, no rescan
```

Avoid calling `get_dataloader()`/`build_dataset()` inside your epoch loop or validator; if you don’t want the late-epoch rebuild, set `close_mosaic=0`. For context on DDP behavior and loader construction see the Multi‑GPU training guide and `build_dataloader`/`InfiniteDataLoader.reset()` reference: [Model Training – Multi‑GPU](https://docs.ultralytics.com/modes/train/), [Data loading reference](https://docs.ultralytics.com/reference/data/build/), and the DDP notes in [Trainer.train](https://docs.ultralytics.com/reference/engine/trainer/).

**Comment by JL-52Hertz** [comment]
This is really a good solution, now my program not only runs successfully but also in the way I expected. Thank you for your replies and help! @glenn-jocher @Y-T-G

**Comment by glenn-jocher** [comment]
Glad it’s working—credit to @Y‑T‑G and the YOLO community; for future tweaks note the validation loader runs with `rect=True` and no shuffle per DetectTrainer’s `get_dataloader()`, see the [detection dataloader reference](https://docs.ultralytics.com/reference/models/yolo/detect/train/#get_dataloader).
